File: Dev/C++/Mascaret/src/BEHAVE/OrganisationalEntity.cpp
#include "BEHAVE/OrganisationalEntity.h"
#include "BEHAVE/RoleAssignement.h"
#include "BEHAVE/RessourceAssignement.h"
#include "BEHAVE/Role.h"
#include "BEHAVE/Ressource.h"
#include "BEHAVE/AgentPlateform.h"
#include "BEHAVE/ACLMessage.h"

namespace BEHAVE
{

OrganisationalEntity::OrganisationalEntity(const string & name) : Element(name)
{}

OrganisationalEntity::~OrganisationalEntity()
{}

shared_ptr<RoleAssignement> OrganisationalEntity::findRoleAssignement(const string& roleName)
{
	for(size_t i=0;i<_roleAssignements.size();i++)
	{
		if(_roleAssignements[i]->getRole()->getName() == roleName)
			return _roleAssignements[i];
	}
	return shared_ptr<RoleAssignement>();
}
shared_ptr<RessourceAssignement> OrganisationalEntity::findRessourceAssignement(const string& resourceName)
{
	for(size_t i=0;i<_ressourceAssignements.size();i++)
	{
		if(_ressourceAssignements[i]->getRessource()->getName() == resourceName)
			return _ressourceAssignements[i];
	}
	return shared_ptr<RessourceAssignement>();
}

void OrganisationalEntity::stopProcedure(const string& procedureName)
{
	for(size_t i=0;i<_roleAssignements.size();i++)
	{
	   shared_ptr<AID> agent = _roleAssignements[i]->getAgent();
	   shared_ptr<ACLMessage> message = shared_ptr<ACLMessage>(new ACLMessage(REQUEST));
           string content = "STOP " + procedureName;
           message->setContent(content);
           message->addReceiver(agent);
           
           MascaretApplication::getInstance()->getAgentPlateform()->sendMessage(message);

	}
}

void OrganisationalEntity::stopAllProcedures()
{
	for(size_t i=0;i<_roleAssignements.size();i++)
	{
	   shared_ptr<AID> agent = _roleAssignements[i]->getAgent();
	   shared_ptr<ACLMessage> message = shared_ptr<ACLMessage>(new ACLMessage(REQUEST));
           string content = "STOP ALL";
           message->setContent(content);
           message->addReceiver(agent);
           
           MascaretApplication::getInstance()->getAgentPlateform()->sendMessage(message);

	}
}

}

File: Dev/CS/Mascaret/Mascaret/BEHAVE/OrganisationalEntity.cs
using System;
using System.Collections.Generic;

namespace Mascaret
{
    public class OrganisationalEntity : Element
    {


        private List<RessourceAssignement> ressourcesAssignement = new List<RessourceAssignement>();
        public List<RessourceAssignement> RessourcesAssignement
        {
            get { return ressourcesAssignement; }
            set { ressourcesAssignement = value; }
        }


        private List<RoleAssignement> roleAssignement = new List<RoleAssignement>();
        public List<RoleAssignement> RoleAssignement
        {
            get { return roleAssignement; }
            set { roleAssignement = value; }
        }


        private OrganisationalStructure structure;
        public OrganisationalStructure Structure
        {
            get { return structure; }
            set { structure = value; }
        }

        public OrganisationalEntity(string name)
            : base(name)
        {
        }

        public RoleAssignement findRoleAssignement(string roleName)
        {
            foreach (RoleAssignement currentRA in roleAssignement)
            {
                if (currentRA.Role.name == roleName)
                    return currentRA;
            }

            return null;
        }

        public RessourceAssignement findRessourceAssignement(string resName)
        {
            
            foreach (RessourceAssignement currentRA in ressourcesAssignement)
            {
                MascaretApplication.Instance.VRComponentFactory.Log("FindRessource :" + currentRA.Ressource.name + " == " + resName);
                if (currentRA.Ressource.name == resName)
                {
                    MascaretApplication.Instance.VRComponentFactory.Log(".... Found");
                    return currentRA;
                }
            }
            return null;
        }

        public void stopProcedure(string procedureName)
        {
            foreach (RoleAssignement currentRA in roleAssignement)
            {
                ACLMessage message = new ACLMessage(ACLPerformative.REQUEST);
                message.Content = "STOP" + procedureName;
                message.Receivers.Add(currentRA.Agent);
                MascaretApplication.Instance.AgentPlateform.sendMessage(message);
            }
        }

        public void stopAllProcedures()
        {
            foreach (RoleAssignement currentRA in roleAssignement)
            {
                ACLMessage message = new ACLMessage(ACLPerformative.REQUEST);
                message.Content = "STOP ALL";
                message.Receivers.Add(currentRA.Agent);
                MascaretApplication.Instance.AgentPlateform.sendMessage(message);
            }
        }

    }
}


